### ISSUE TITLE
old versions

### ISSUE THREAD

**Issue by timosachsenberg** [OpenMS/OpenMS-docs#231]
![image](https://github.com/OpenMS/OpenMS-docs/assets/5803621/b743bd83-af26-49ba-baaa-5df412bc3f3d)

**Comment by timosachsenberg** [comment]
on abibuilder ![image](https://github.com/OpenMS/OpenMS-docs/assets/5803621/e969342f-af15-4a40-a003-9d23fcd3d8bd)

**Comment by timosachsenberg** [comment]
@jpfeuffer or @poshul or @matteopilz do you know why these are not displayed? they seem to be in the abibuilder folders

**Comment by timosachsenberg** [comment]
seems to me as if the nightly documentation can't link to the release documentation. anyone can come up with an easy fix for that?

**Comment by KyowonJeong** [comment]
It is broken for both versions. Only index.html is working but subpages are broken.

**Comment by jpfeuffer** [comment]
I thought ~~The selector only works from openms.de/org, not from the archive.
It cannot be adapted to handle both "path" hierarchies.~~

**Comment by jpfeuffer** [comment]
But even there, it is true that it apparently only works on the index.html. I have no idea why.

**Comment by jpfeuffer** [comment]
<img width="1098" alt="Screenshot 2024-06-19 at 15 31 40" src="https://github.com/OpenMS/OpenMS-docs/assets/8102638/dfcc7b64-64da-48be-b776-ccde74183ef6">

It just checks the wrong links. For everything but index.html it should add a html subfolder.
Code is here:
https://github.com/OpenMS/OpenMS/blob/develop/doc/doxygen/common/selectversion.js#L24

**Comment by JKVISION1101** [comment]
I checked this problem. 

I think index.html links to the correct site, but subpages(related pages, modules, classes...etc) links to incorrect directory, causing problems.

For example.

if i click 2.8.0 version in index.html, it is linked to this page 

`https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.8.0/html/index.html`


However, 

`https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.8.0/pages.html` was linked to subpages version selector.


The difference between index.html and subpages is that **/html/** is missing from subpages..

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/340348db-5fdf-40c2-a7ca-0bb7223a5eff)




If i add `/html/` after 2.8.0 in the subpages, it works very well.

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/fa0497b0-c495-454f-8f28-d6e5a45c1a17)




I think this problem can be easily resolved by adding /html/ after version in the directory like index.html.

I want to delve deeper into this issue, but I don't know how to build the OpenMS Doxygen.


 If you can tell me how to build openms doxygen, I will be able to modify the code.

**Comment by jpfeuffer** [comment]
> <img width="1098" alt="Screenshot 2024-06-19 at 15 31 40" src="https://github.com/OpenMS/OpenMS-docs/assets/8102638/dfcc7b64-64da-48be-b776-ccde74183ef6">
> 
> It just checks the wrong links. For everything but index.html it should add a html subfolder.
> Code is here:
> https://github.com/OpenMS/OpenMS/blob/develop/doc/doxygen/common/selectversion.js#L24

@JKVISION1101 Thanks for checking. You are correct with the evaluation of the problem but I think it is easier to fix it programmatically in the redirecting code linked above (instead of moving the folder structures).

Should be a simple

```
if not basename == index.html:
  redirectfolder.append("html").append(basename)
```

(Pseudocode)

Up to you.

**Comment by JKVISION1101** [comment]
@jpfeuffer 

Thanks for reply. 

Your comment helped a lot.

I built doc in the [openms build foler]/doc/OpenMS_doc.sln successfully.

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/4522ac90-dffb-4139-bd9c-dd6fbe917a3f)




However, the output index.html has button for version selector, but doesn't have dropdown for version.

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/913b00c2-a8a7-46f1-b00a-e8987bea9585)

Do you know how to make dropdown come out?

**Comment by jpfeuffer** [comment]
You will need to build another version and put it in the same parent folder.

**Comment by JKVISION1101** [comment]
@jpfeuffer 

I checked selectversion.js file.

`    docs = docs.map((x) => '<a class="verslink" href="'
        + urlrootdir
        + '/release/' + x + '/' + patharr[patharr.length - 1] + '">'
        + x
        + '</a>');`


i think this code is related with site directory linked to specifiic version.

and there are only 

/archive/openms/Documentation/release/**version**/index.html



How about add /html/ in this code?


`    docs = docs.map((x) => '<a class="verslink" href="'
        + urlrootdir
        + '/release/' + x + '/' + 'html/' + patharr[patharr.length - 1] + '">'
        + x
        + '</a>');`


However, one thing I don't understand is why the index page and the subfolder have the same HTML header, but the generated links are different.

**Comment by JKVISION1101** [comment]
@jpfeuffer 

Could you tell me how can i build anoter version and put same parent folder in detail.

and then I want to try above code.


Another suspicious aspect is that the fact that the header file is not being applied in the doxyfile_dot.in file is also questionable.


![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/7ade8dfa-3d8a-42fd-93d4-602844a977bf)

Could you tell me what is doxyfile_dot.in's role?

**Comment by jpfeuffer** [comment]
You can just copy the docs of same version, rename and put them next to each other.
doxyfile_dot is not used for the full documentation.

**Comment by jpfeuffer** [comment]
> @jpfeuffer 
> 
> I checked selectversion.js file.
> 
> `    docs = docs.map((x) => '<a class="verslink" href="'
>         + urlrootdir
>         + '/release/' + x + '/' + patharr[patharr.length - 1] + '">'
>         + x
>         + '</a>');`
> 
> 
> i think this code is related with site directory linked to specifiic version.
> 
> and there are only 
> 
> /archive/openms/Documentation/release/**version**/index.html
> 
> 
> 
> How about add /html/ in this code?
> 
> 
> `    docs = docs.map((x) => '<a class="verslink" href="'
>         + urlrootdir
>         + '/release/' + x + '/' + 'html/' + patharr[patharr.length - 1] + '">'
>         + x
>         + '</a>');`


Yes this is the solution. But you need an if-case for the index.html. For index.html, the current code is correct.


> However, one thing I don't understand is why the index page and the subfolder have the same HTML header, but the generated links are different.

What do you mean the links are different? I think they are the same and this is the problem.

**Comment by JKVISION1101** [comment]
>What do you mean the links are different? I think they are the same and this is the problem.


The index page doesn't produce an error because the linked address contains "html."
(I am not sure as  abibuilder is not accessible now.)

However, the subfolder page does produce an error because its linked address doesn't contain "html." That's why I mentioned that the generated links are different. 

Both the index and subfolder pages apply the same selectversion.js. I'm curious why the index page includes "html" in the address and doesn't encounter an error, while the subfolder page doesn't include "html" and does encounter an error.



Currently, abibuilder is not accessible, so my understanding might be incorrect. Please bear with me.

**Comment by JKVISION1101** [comment]
@jpfeuffer 
> You can just copy the docs of same version, rename and put them next to each other.
doxyfile_dot is not used for the full documentation.


in  my case,  i copied doc folder and rename "2.8.0" in the "openms_development folder""
![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/9d8991d2-cd27-42c2-9892-af8121451265)




And then go to the doc file and build doc target in the openms_doc.sln.
![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/342bb3b6-20be-4779-94bd-e5b2a0e65ece)

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/dd6437f5-8e49-48fc-adf1-2bbb3fc657ec)

However, there is no dropdown in the index.html file in the doc folder.

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/fdd10ea9-1503-4fc3-836c-64f7bf7bcfaa)




I'm always grateful for your answers.

**Comment by jpfeuffer** [comment]
> >What do you mean the links are different? I think they are the same and this is the problem.
> 
> 
> The index page doesn't produce an error because the linked address contains "html."
> (I am not sure as  abibuilder is not accessible now.)
> 
> However, the subfolder page does produce an error because its linked address doesn't contain "html." That's why I mentioned that the generated links are different. 
> 
> Both the index and subfolder pages apply the same selectversion.js. I'm curious why the index page includes "html" in the address and doesn't encounter an error, while the subfolder page doesn't include "html" and does encounter an error.
> 
> 
> 
> Currently, abibuilder is not accessible, so my understanding might be incorrect. Please bear with me.


No, it is the other way round. Selectversion.js NEVER adds html. Index.html works WITHOUT "html/", the rest of the pages only works WITH "html/". It all depends on where the pages lie on the webserver. It is common practice to have the homepage under folder/index.html and the rest under folder/html/anysubpage.html

**Comment by jpfeuffer** [comment]
> @jpfeuffer 
> > You can just copy the docs of same version, rename and put them next to each other.
> doxyfile_dot is not used for the full documentation.
> 
> 
> in  my case,  i copied doc folder and rename "2.8.0" in the "openms_development folder""
> ![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/9d8991d2-cd27-42c2-9892-af8121451265)
> 
> 
> 
> 
> And then go to the doc file and build doc target in the openms_doc.sln.
> ![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/342bb3b6-20be-4779-94bd-e5b2a0e65ece)
> 
> ![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/dd6437f5-8e49-48fc-adf1-2bbb3fc657ec)
> 
> However, there is no dropdown in the index.html file in the doc folder.
> 
> ![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/fdd10ea9-1503-4fc3-836c-64f7bf7bcfaa)
> 
> 
> 
> 
> I'm always grateful for your answers.

Selectversion.js needs very specific folder structures/paths that are adapted to our abibuilder/openms.de

https://github.com/OpenMS/OpenMS/blob/27a46abba2dedf9ebc9aeb826074faa030297a09/doc/doxygen/common/selectversion.js#L2

You need to either adapt it to be more general or reproduce the exact folder structure locally.

**Comment by jpfeuffer** [comment]
You can also just adapt selectversion.js to your local structure for testing, or if possible even better, make it more general overall to work with more structures.

You can also fix selectversion.js by hard coding/mocking a set of paths (e.g. by copying them from the browser and pasting them in an array). Then comment out the parseDirectoryListing function (and revert before commiting).

**Comment by JKVISION1101** [comment]
@jpfeuffer 

Thanks Julianus.

I think I found the solution.


This is index page. and as seen below, the href don't have html. 

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/deb0f20a-714e-46b4-b6a8-6f7a76724f66)


And the rest of pages also don't have html

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/1bcc6d7a-5058-4a0f-8b08-f1fccdc7d49e)




So, selectversion.js doesn't make 'html' in both index page and the rest of pages.

<br/>
<br/>
<br/>
<br/>
<br/>














the reason why only index page works is that  even if the site address does not include 'html', accessing the site automatically generates 'html'.

for example, In index page, If i click red one (this href is "archive/openms/Documentation/release/2.8.0/index.html") in the below image, 

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/2912f228-6d3f-4f3f-80e7-ebefb2878450)


however, this go to https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.8.0/html/index.html (including "html")

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/4b17a96c-94b8-447f-bae9-7d2a7f5da450)




<br/>
<br/>
<br/>
<br/>
<br/>

However,  In rest of pages, If i clicked red one (this href is https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.8.0/pages.html) in the below image,

![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/860b17f6-9a12-473f-a0c1-2d3273cc8446)

however, this go to https://abibuilder.cs.uni-tuebingen.de/archive/openms/Documentation/release/2.8.0/index.html ( NOT including "html") and this make error.
![image](https://github.com/OpenMS/OpenMS-docs/assets/120628409/0ce07820-9f1d-4f00-8cbd-4d666634f394)




In summary, It operates solely based on the index page, which automatically redirects to where 'html' exists in the site directory, even if 'html' is not explicitly included. (So, If we add html in the index page by changing selectversion.js code, it doesn't make problem)






`    docs = docs.map((x) => '<a class="verslink" href="' + urlrootdir + '/release/' + x + '/' + 'html/' + patharr[patharr.length - 1] + '">' + x + '</a>');`


Changing the code as described above will resolve the issue.

**Comment by jpfeuffer** [comment]
Yes, true. Since root/index.html redirects to root/html/index.html, I guess **always** adding html should be fine!
Feel free to prepare a pull request.

**Comment by JKVISION1101** [comment]
@jpfeuffer 

I tried pull request adding html in the selectversion.js.

However, I haven't do the pull request. so I am not sure that i did it properly.

If possible, please check it.

I really appreciated your help.

Thanks.
